_posts/2019-06-07-python-datascience-with-openfaas.md

```diff
@@ -0,0 +1,267 @@
+
+```
+function
+├── __init__.py
+├── core
+│   ├── README.md
+│   ├── __init__.py
+│   └── utils.py
+├── handler.py
+└── train.py
+```
+
+Like the basic `python` template, only the `handler.py` implementation is required, and both the `core` module and the `train.py` could be replaced if you needed, but this structure starts you on a path to keeping the code organized.
+
+## Name classifier implementation
+
+### From a template to actual code
+The name classifier function uses the [neural network implementation in PyTorch][pytorch-nn-docs].  PyTorch has a great [introduction and walk-through][pytorch-nn-intro] for neural network package and [model training][pytorch-training]. In this post we focus on what a concrete and complete implementation looks like. All of the interesting details about the [implementation of the model][classifier-impl], the utilities used to parse the training data, and the utility for applying the model to a new inputs are stored in the `core` module.  This `core` module can then be used in _both_ the function handler and the training script. This code is also easier to reuse in a new function because it is cleanly isolated from the function itself.
+
+```
+classify
+├── __init__.py
+├── core
+│   ├── README.md
+│   ├── __init__.py
+│   ├── const.py
+│   ├── model.py
+│   └── utils.py
+├── data
+│   ├── char-rnn-classification.pt
+│   └── names
+│       ├── Arabic.txt
+│       ├── Chinese.txt
+│       ├── ...
+│       └── Vietnamese.txt
+├── handler.py
+├── requirements.txt
+└── train.py
+```
+
+### Training
+Another point to note is the training data folder is also included here, `data/names` and a serialized model is also include `data/char-rnn-classification.pt`.
+
+This template is designed to run the training as part of the build process.  You can easily trigger the training using
+
+```bash
+python train.py
+```
+
+This will generate a new serialized model `data/char-rnn-classification.pt`.
+
+There are two methods for [saving PyTorch models][pytorch-saving-models]:
+
+1. save the entire model using `pickle`, or
+2. saving just the model `state_dict`.
+
+
+We have opted for the second method because it is more portable, the `pickle` method is sensitive to the Python environment and folder structure.
+
+```python
+torch.save(rnn.state_dict(), "data/char-rnn-classification.pt")
+```
+
+### Loading the model into function memory
+Using the `HTTP` mode in the [of-watchdog][of-watchdog-homepage] enables us to load this model into memory once and reuse it for multiple requests. We only need to load the model at the start of the `handler.py` implementation
+
+```python
+import json
+import os
+from pathlib import PurePath
+from typing import Any, List
+
+import torch
+from torch.autograd import Variable
+
+from .core import const, model, utils
+
+FUNCTION_ROOT = os.environ.get("function_root", "/root/function/")
```

Comment:
`/root/function/` ? I thought we were non-root?

Reply:
that is just the name of the folder, not the user https://github.com/LucasRoesler/pydatascience-template/blob/master/template/pydatatscience/Dockerfile

Reply:
https://github.com/openfaas/templates/blob/master/template/python3/Dockerfile#L32